File: formatter.go
// Fills a token budget with the highest-ranked evidence, and states plainly
// what it did not show. Counters are always the true ones; nothing is silently
// dropped and no synthetic marker is ever injected into source text.

package main

import (
	"fmt"
	"math"
	"strings"
)

const displayMax = 150

// Code sits around 3.3 characters per token; bytes/4 (what some tools use)
// overestimates savings on source lines.
const charsPerToken = float32(3.3)

func estTokens(s string) int {
	return int(math.Round(float64(float32(runeLen(s)) / charsPerToken)))
}

func clip(s string) string {
	out, cut := takeRunes(s, displayMax)
	if cut {
		return out + "…"
	}
	return out
}

func entry(f *FileHit, explains map[string]Explain) string {
	var l *Line
	if f.Def != nil {
		l = f.Def
	} else if len(f.Best) > 0 {
		l = &f.Best[0]
	}
	var s strings.Builder
	if l == nil {
		fmt.Fprintf(&s, "%s  (%dx)\n", f.Path, f.Matches)
		return s.String()
	}
	fmt.Fprintf(&s, "%s:%d", f.Path, l.N)
	if f.Matches > 1 {
		fmt.Fprintf(&s, "  (%dx)", f.Matches)
	}
	if l.IsDef {
		s.WriteString("  [def]")
	}
	if explains != nil {
		if e, ok := explains[f.Path]; ok {
			fmt.Fprintf(&s, "  {score=%.2f line=%.2f cov=%.2f dens=%.2f path=%.2f name=%.2f dedup=%.2f}",
				f.Score, e.Line, e.Coverage, e.Density, e.PathPrior, e.NameBonus, e.Dedup)
		}
	}
	s.WriteString("\n    ")
	s.WriteString(clip(l.Text))
	s.WriteString("\n")
	return s.String()
}

type RenderOpts struct {
	Query string
	// The literal query matched nothing and we fell back to term coverage.
	Fallback     bool
	BudgetTokens int
	Page         int
	Repro        string
	ElapsedMs    int64
	Explains     map[string]Explain
}

func render(out *Outcome, opts RenderOpts) string {
	if out.TotalMatch == 0 {
		return fmt.Sprintf("query: %s   hits=0 files=0\nrepro: %s\n", opts.Query, opts.Repro)
	}

	// Header and footer are part of the budget, so reserve for them.
	budgetChars := int(float32(opts.BudgetTokens) * charsPerToken)
	usable := budgetChars - 260
	if usable < 200 {
		usable = 200
	}

	// Walk the ranked list page by page; a page is full when the next entry
	// would cross the budget. Deterministic, so --page N is stable.
	type rng struct{ from, to int }
	var pages []rng
	start, used := 0, 0
	for i, f := range out.Files {
		n := runeLen(entry(f, opts.Explains))
		if used+n > usable && i > start {
			pages = append(pages, rng{start, i})
			start = i
			used = 0
		}
		used += n
	}
	pages = append(pages, rng{start, len(out.Files)})

	totalPages := len(pages)
	page := opts.Page
	if page < 1 {
		page = 1
	}
	if page > totalPages {
		page = totalPages
	}
	from, to := pages[page-1].from, pages[page-1].to

	var body strings.Builder
	for _, f := range out.Files[from:to] {
		body.WriteString(entry(f, opts.Explains))
	}

	shown := to - from
	remaining := len(out.Files) - to
	// Definitions are sorted to the front, so any that are off-page are on a
	// later page — the count is stated rather than left implicit.
	defsOffpage := 0
	for i, f := range out.Files {
		if (i < from || i >= to) && f.Def != nil {
			defsOffpage++
		}
	}

	var tail strings.Builder
	tail.WriteString("\n")
	if remaining > 0 {
		fmt.Fprintf(&tail,
			"+%d files not shown — scout search '%s' --page %d   (or --path/--lang/--kind symbol to narrow)\n",
			remaining, opts.Query, page+1)
	}
	if defsOffpage > 0 {
		fmt.Fprintf(&tail, "note: %d file(s) holding a definition are not on this page\n", defsOffpage)
	}
	fmt.Fprintf(&tail, "repro: %s\n", opts.Repro)

	warn := ""
	if opts.Fallback {
		// No indent on the second line: the reference uses a Rust line
		// continuation, which strips the leading whitespace of the wrapped line.
		warn = "note: no match for the literal string. Fell back to term coverage —\n" +
			"ranking is unreliable here, prefer an exact identifier or --kind symbol.\n"
	}
	header := fmt.Sprintf("query: %s   hits=%d files=%d shown=%d page=%d/%d   ~%dtok  %dms\n%s\n",
		opts.Query, out.TotalMatch, out.TotalFiles, shown, page, totalPages,
		estTokens(body.String())+estTokens(tail.String()), opts.ElapsedMs, warn)

	return header + body.String() + tail.String()
}

File: src/formatter.rs
//! Fills a token budget with the highest-ranked evidence, and states plainly
//! what it did not show. Counters are always the true ones; nothing is
//! silently dropped and no synthetic marker is ever injected into source text.

use std::collections::HashMap;

use crate::executor::{FileHit, Outcome};
use crate::ranker::Explain;

const DISPLAY_MAX: usize = 150;
/// Code sits around 3.3 characters per token; `bytes / 4` (what rtk uses)
/// overestimates savings on source lines.
const CHARS_PER_TOKEN: f32 = 3.3;

pub fn est_tokens(s: &str) -> usize {
    (s.chars().count() as f32 / CHARS_PER_TOKEN).round() as usize
}

fn clip(s: &str) -> String {
    if s.chars().count() <= DISPLAY_MAX {
        return s.to_string();
    }
    let mut out: String = s.chars().take(DISPLAY_MAX).collect();
    out.push('…');
    out
}

fn entry(f: &FileHit, explains: Option<&HashMap<String, Explain>>) -> String {
    let l = f.def.as_ref().or_else(|| f.best.first());
    let mut s = String::new();
    match l {
        Some(l) => {
            s.push_str(&format!("{}:{}", f.path, l.n));
            if f.matches > 1 {
                s.push_str(&format!("  ({}x)", f.matches));
            }
            if l.is_def {
                s.push_str("  [def]");
            }
            if let Some(e) = explains.and_then(|m| m.get(&f.path)) {
                s.push_str(&format!(
                    "  {{score={:.2} line={:.2} cov={:.2} dens={:.2} path={:.2} name={:.2} dedup={:.2}}}",
                    f.score, e.line, e.coverage, e.density, e.path_prior, e.name_bonus, e.dedup
                ));
            }
            s.push('\n');
            s.push_str("    ");
            s.push_str(&clip(&l.text));
            s.push('\n');
        }
        None => {
            s.push_str(&format!("{}  ({}x)\n", f.path, f.matches));
        }
    }
    s
}

pub struct RenderOpts<'a> {
    pub query: &'a str,
    /// The literal query matched nothing and we fell back to term coverage.
    pub fallback: bool,
    pub budget_tokens: usize,
    pub page: usize,
    pub repro: String,
    pub elapsed_ms: u128,
    pub explains: Option<&'a HashMap<String, Explain>>,
}

pub fn render(out: &Outcome, opts: RenderOpts<'_>) -> String {
    if out.total_matches == 0 {
        return format!(
            "query: {}   hits=0 files=0\nrepro: {}\n",
            opts.query, opts.repro
        );
    }

    // Header and footer are part of the budget, so reserve for them.
    let budget_chars = (opts.budget_tokens as f32 * CHARS_PER_TOKEN) as usize;
    let reserve = 260usize;
    let usable = budget_chars.saturating_sub(reserve).max(200);

    // Walk the ranked list page by page; a page is full when the next entry
    // would cross the budget. Deterministic, so `--page N` is stable.
    let mut pages: Vec<(usize, usize)> = Vec::new();
    let mut start = 0usize;
    let mut used = 0usize;
    for (i, f) in out.files.iter().enumerate() {
        let e = entry(f, opts.explains);
        let n = e.chars().count();
        if used + n > usable && i > start {
            pages.push((start, i));
            start = i;
            used = 0;
        }
        used += n;
    }
    pages.push((start, out.files.len()));

    let total_pages = pages.len();
    let page = opts.page.clamp(1, total_pages);
    let (from, to) = pages[page - 1];

    let mut body = String::new();
    for f in &out.files[from..to] {
        body.push_str(&entry(f, opts.explains));
    }

    let shown = to - from;
    let remaining = out.files.len() - to;
    // Definitions are sorted to the front, so any that are off-page are on a
    // later page — the count is stated rather than left implicit.
    let defs_offpage = out.files[..from]
        .iter()
        .chain(out.files[to..].iter())
        .filter(|f| f.def.is_some())
        .count();

    let mut tail = String::from("\n");
    if remaining > 0 {
        tail.push_str(&format!(
            "+{} files not shown — scout search '{}' --page {}   (or --path/--lang/--kind symbol to narrow)\n",
            remaining,
            opts.query,
            page + 1
        ));
    }
    if defs_offpage > 0 {
        tail.push_str(&format!(
            "note: {} file(s) holding a definition are not on this page\n",
            defs_offpage
        ));
    }
    tail.push_str(&format!("repro: {}\n", opts.repro));

    let warn = if opts.fallback {
        "note: no match for the literal string. Fell back to term coverage —\n\
         ranking is unreliable here, prefer an exact identifier or --kind symbol.\n"
    } else {
        ""
    };
    let header = format!(
        "query: {}   hits={} files={} shown={} page={}/{}   ~{}tok  {}ms\n{}\n",
        opts.query,
        out.total_matches,
        out.total_files,
        shown,
        page,
        total_pages,
        est_tokens(&body) + est_tokens(&tail),
        opts.elapsed_ms,
        warn
    );
    header + &body + &tail
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tokens_estimate() {
        assert_eq!(est_tokens(""), 0);
        assert!(est_tokens(&"x".repeat(33)) == 10);
    }

    #[test]
    fn clip_marks_truncation_without_inventing_syntax() {
        let long = "a".repeat(400);
        let c = clip(&long);
        assert!(c.ends_with('…'));
        assert!(!c.contains("//"), "no synthetic comment is ever injected");
    }
}
